src/moin/storage/middleware/protecting.py

```diff
@@ -162,36 +177,65 @@ def search_meta(self, q, idx_name=LATEST_REVS, **kw):
         The intended use of this method is to return the current rev metadata for all
         of the items in namespace subject to query restrictions. This is useful for reports
         such as Global Index, Global Tags, Wanted Items, Orphaned Items, etc.
-
-        Save processing time by avoiding a full ACL check when the answer will be the same as the last.
         """
+        from . indexing import parent_names
+
         last_rev_acl_parts = (None, None, None)
         last_rev_result = None
-        for rev in self.indexer.search_meta(q, idx_name, **kw):
-            rev = ProtectedItemMeta(self, rev)
-            this_rev_acl_parts = (rev.meta[NAMESPACE], rev.meta.get(PARENTNAMES), rev.meta.get(ACL))
-            if this_rev_acl_parts == last_rev_acl_parts:
-                # skip the acl check because we know the answer will be the same
-                if last_rev_result:
-                    yield rev
-                continue
-            last_rev_acl_parts = this_rev_acl_parts
-            result = rev.allows(READ) or rev.allows(PUBREAD)
-            last_rev_result = result
+        for meta in self.indexer.search_meta(q, idx_name, **kw):
+            self.meta = meta
+            meta[FQNAMES] = gen_fqnames(meta)
+            self.fqnames = meta[FQNAMES]
+            result = self.allows(tuple(self.user.name), meta.get(ACL, None), tuple(parent_names(meta[NAME])), meta[NAMESPACE], READ)
             if result:
-                yield rev
+                yield meta
 
-    def may_read_rev(self, windex):
+    def may_read_rev(self, meta):
         """
         Return true if user may read item revision represented by whoosh index hit.
-        Called by search templates, maybe others.
+        Called by ajaxsearch template.
         """
-        hit = ProtectedItemMeta(self, windex)
-        may_read = hit.allows(READ) or hit.allows(PUBREAD)
+        from . indexing import parent_names
```

Comment:
same here

Reply:
also this method seems to have duplicate code as already used above, so maybe above code could just call this method?

Reply:
will fix